org.lflang/src/org/lflang/generator/ts/TSReactionGenerator.kt

```diff
@@ -0,0 +1,345 @@
+        return if (v.container != null) {
+            "__${v.container.name}_${v.variable.name}"
+        } else {
+            "__${v.variable.name}"
+        }
+    }
+
+    private fun generateDeadlineHandler(
+        reaction: Reaction,
+        reactPrologue: String,
+        reactEpilogue: String,
+        reactSignature: StringJoiner
+    ): String {
+        var deadlineArgs = ""
+        if (reaction.deadline.delay.parameter != null) {
+            deadlineArgs += "this." + reaction.deadline.delay.parameter.name + ".get()";
+        } else {
+            deadlineArgs += reaction.deadline.delay.getTargetValue()
+        }
+
+        return with(PrependOperator) {
+            """
+            |},
+            |$deadlineArgs,
+            |function($reactSignature) {
+            |    // =============== START deadline prologue
+        ${" |    "..reactPrologue}
+            |    // =============== END deadline prologue
+            |    try {
+        ${" |        "..reaction.deadline.code.toText()}
+            |    } finally {
+            |        // =============== START deadline epilogue
+        ${" |        "..reactEpilogue}
+            |        // =============== END deadline epilogue
+            |    }
+            |}
+        """.trimMargin()
+        }
+
+    }
+
+    private fun generateReactionString(
+        reaction: Reaction,
+        reactPrologue: String,
+        reactEpilogue: String,
+        reactFunctArgs: StringJoiner,
+        reactSignature: StringJoiner
+    ): String {
+        // Assemble reaction triggers
+        val reactionTriggers = StringJoiner(",\n")
+        for (trigger in reaction.triggers) {
+            if (trigger is VarRef) {
+                reactionTriggers.add("this." + trigger.generateVarRef())
+            } else if (trigger.isStartup) {
+                reactionTriggers.add("this.startup")
+            } else if (trigger.isShutdown) {
+                reactionTriggers.add("this.shutdown")
+            }
+        }
+        return with(PrependOperator) {
+            """
+            |
+            |this.addReaction(
+            |    new __Triggers($reactionTriggers),
+            |    new __Args($reactFunctArgs),
+            |    function ($reactSignature) {
+            |        // =============== START react prologue
+        ${" |        "..reactPrologue}
+            |        // =============== END react prologue
+            |        try {
+        ${" |            "..reaction.code.toText()}
+            |        } finally {
+            |            // =============== START react epilogue
+        ${" |            "..reactEpilogue}            
+            |            // =============== END react epilogue
+            |        }
+        ${" |    "..if (reaction.deadline != null) generateDeadlineHandler(reaction, reactPrologue, reactEpilogue, reactSignature) else "}"}
+            |);
+        """.trimMargin()
+        }
+    }
+
+    fun generateReaction(
```

Comment:
This can definitely be decomposed further, but it's a good start to have factored this stuff out of `TSGenerator`.

Reply:
Agreed, I added a TODO to make sure to break this function into multiple functions.